From: Benjamin Tissoires <benjamin.tissoires@gmail.com>
Date: Wed, 14 Nov 2012 16:59:16 +0100
Subject: HID: add usage_index in struct hid_usage.
Git-commit: f262d1fa2c651a5e2f92b6aee8779597631cd5d4
Patch-mainline: v3.8-rc1
References: bnc#835839

Currently, there is no way to know the index of the current field
in the .input_mapping and .event callbacks  when this field is inside
an array of HID fields.
This patch adds this index to the struct hid_usage so that this
information is available to input_mapping and event callbacks.

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@gmail.com>
Acked-by: Jiri Kosina <jkosina@suse.cz>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
Acked-by: Borislav Petkov <bp@suse.de>
---
 drivers/hid/hid-core.c | 4 ++++
 include/linux/hid.h    | 1 +
 2 files changed, 5 insertions(+)

Index: current/drivers/hid/hid-core.c
===================================================================
--- current.orig/drivers/hid/hid-core.c	2014-10-29 18:08:21.348377522 +0100
+++ current/drivers/hid/hid-core.c	2014-10-29 18:11:14.644377215 +0100
@@ -94,6 +94,7 @@ EXPORT_SYMBOL_GPL(hid_register_report);
 static struct hid_field *hid_register_field(struct hid_report *report, unsigned usages, unsigned values)
 {
 	struct hid_field *field;
+	int i;
 
 	if (report->maxfield == HID_MAX_FIELDS) {
 		hid_err(report->device, "too many fields in report\n");
@@ -112,6 +113,9 @@ static struct hid_field *hid_register_fi
 	field->value = (s32 *)(field->usage + usages);
 	field->report = report;
 
+	for (i = 0; i < usages; i++)
+		field->usage[i].usage_index = i;
+
 	return field;
 }
 
Index: current/include/linux/hid.h
===================================================================
--- current.orig/include/linux/hid.h	2014-10-29 18:08:21.348377522 +0100
+++ current/include/linux/hid.h	2014-10-29 18:11:14.644377215 +0100
@@ -342,6 +342,7 @@ struct hid_collection {
 struct hid_usage {
 	unsigned  hid;			/* hid usage code */
 	unsigned  collection_index;	/* index into collection array */
+	unsigned  usage_index;		/* index into usage array */
 	/* hidinput data */
 	__u16     code;			/* input driver code */
 	__u8      type;			/* input driver type */
